nexus/src/app/instance.rs

```diff
@@ -1394,33 +1394,61 @@ impl super::Nexus {
 
     pub(crate) async fn instance_serial_console_stream(
         &self,
-        conn: dropshot::WebsocketConnection,
+        mut client_stream: WebSocketStream<impl AsyncRead + AsyncWrite + Unpin>,
         instance_lookup: &lookup::Instance<'_>,
         params: &params::InstanceSerialConsoleStreamRequest,
     ) -> Result<(), Error> {
-        let client = self
+        let client = match self
             .propolis_client_for_instance(
                 instance_lookup,
                 authz::Action::Modify,
             )
-            .await?;
+            .await
+        {
+            Ok(x) => x,
+            Err(e) => {
+                let _ = client_stream
+                    .close(Some(CloseFrame {
+                        code: CloseCode::Error,
+                        reason: e.to_string().into(),
+                    }))
+                    .await
+                    .is_ok();
+                return Err(e);
+            }
+        };
+
         let mut req = client.instance_serial();
         if let Some(most_recent) = params.most_recent {
             req = req.most_recent(most_recent);
         }
-        let propolis_upgraded = req
-            .send()
-            .await
-            .map_err(|_| {
-                Error::internal_error(
-                    "failed to connect to instance's propolis server",
+        match req.send().await {
+            Ok(response) => {
+                let propolis_upgraded = response.into_inner();
+                let log = self.log.clone();
+                Self::proxy_instance_serial_ws(
+                    client_stream,
+                    propolis_upgraded,
+                    Some(log),
                 )
-            })?
-            .into_inner();
-
-        Self::proxy_instance_serial_ws(conn.into_inner(), propolis_upgraded)
-            .await
-            .map_err(|e| Error::internal_error(&format!("{}", e)))
+                .await
+                .map_err(|e| Error::internal_error(&format!("{}", e)))
+            }
+            Err(e) => {
+                let message = format!(
+                    "Failed to connect to instance's propolis server: {}",
```

Comment:
This predates this change and is present elsewhere, and so is very nitpicky (and could also be deferred to an issue), but I wonder if we should also take this opportunity to remove references to "Propolis" from the error strings in this path, especially if we're being very explicit about sending these strings back as reason strings in a websocket close message. (If I didn't know what Propolis was I might be confused by the message string.) I like the "websocket connection to client closed unexpectedly" verbiage below as a substitute; "websocket connection to instance's serial port" also has a nice ring to it, I think.

Again, totally optional for this PR, but might be a bit of fit-and-finish we want to file an issue for and come back to.

Reply:
agreed, just tossed it in here before merge: https://github.com/oxidecomputer/omicron/pull/2871/commits/db1e8f9ba9362731e3f6e89ec579126fd1ddcfae